### PR TITLE
refactor(optimizer): record error contexts when casting composite types

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Record contexts for error when recursively calling `cast` on composite types, to provide better error messages.

Preview:

https://github.com/risingwavelabs/risingwave/blob/fc4223348295f029a4d88d0d9f64b4de8a760ac8/src/frontend/planner_test/tests/testdata/output/cast.yaml#L92-L115

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/output/cast.yaml [hunk 1]
```diff
@@ -80,3 +80,35 @@
...
+    create table t (v map(int, int));
+    select v::map(int, bytea) from t;
+  binder_error: |
+    Failed to bind expression: CAST(v AS MAP(INT,BYTEA))
+
+    Caused by these errors (recent errors listed first):
+      1: cannot cast type "map(integer,integer)" to "map(integer,bytea)" in Explicit context
+      2: cannot cast type "struct<key integer, value integer>[]" to "struct<key integer, value bytea>[]" in Explicit context
+      3: cannot cast type "struct<key integer, value integer>" to "struct<key integer, value bytea>" in Explicit context
+      4: cannot cast type "integer" to "bytea" in Explicit context
```
Comment (xxchan) [inline src/frontend/planner_test/tests/testdata/output/cast.yaml:114]: This might be a little confusing to users 🤪

Comment (BugenZhao) [reply src/frontend/planner_test/tests/testdata/output/cast.yaml]: Updated. Check check again.

File: src/frontend/src/expr/type_inference/cast.rs [hunk 2]
```diff
@@ -204,10 +275,10 @@ pub enum CastContext {
     Explicit,
 }
 
-pub type CastMap = BTreeMap<(DataTypeName, DataTypeName), CastContext>;
+pub type CastTable = BTreeMap<(DataTypeName, DataTypeName), CastContext>;
```
Comment (BugenZhao) [inline src/frontend/src/expr/type_inference/cast.rs:278]: Not to be confused with `cast_map`.

File: src/frontend/src/expr/type_inference/cast.rs [hunk 3]
```diff
@@ -114,76 +116,145 @@ pub fn align_array_and_element(
     Ok(array_type)
 }
 
+/// Returns `Ok` if `ok` is true, otherwise returns a placeholder [`CastError`] to be further
+/// wrapped with a more informative context in [`cast`].
+fn canmeh(ok: bool) -> CastResult {
+    if ok {
+        Ok(())
+    } else {
+        bail_cast_error!()
```
Comment (xxchan) [inline src/frontend/src/expr/type_inference/cast.rs:125]: A little confused how it works. So this is an empty leaf CastError? Why it's not shown in the err msg stack? (Why we need it?)

File: src/frontend/src/expr/type_inference/cast.rs [hunk 4]
```diff
@@ -114,76 +116,145 @@ pub fn align_array_and_element(
     Ok(array_type)
 }
 
+/// Returns `Ok` if `ok` is true, otherwise returns a placeholder [`CastError`] to be further
+/// wrapped with a more informative context in [`cast`].
+fn canmeh(ok: bool) -> CastResult {
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/frontend/src/expr/type_inference/cast.rs]: Sure, renamed.

**Review by xxchan — APPROVED** [review]
very good 👍

File: src/frontend/src/expr/type_inference/cast.rs (hunk 3 above)
Comment (BugenZhao) [reply src/frontend/src/expr/type_inference/cast.rs:125]: Yes. In error report we'll disregard all empty and repeated entries, so it won't appear.

To preserve the reason for a cast failure, we must return a `Result` in functions like `cast_struct`. However, the default error message "cannot cast" is uniformly applied in `cast`, so we simply return an empty leaf error and let the caller to fill the context.

Comment (xxchan) [reply src/frontend/src/expr/type_inference/cast.rs:125]: It sounds like `Result<(), Option<CastError>>` 🤪

**Pull request merged**